qadence/ml_tools/callbacks/writer_registry.py

```diff
@@ -0,0 +1,375 @@
+            result (OptimizeResult): The optimization results to log.
+        """
+        raise NotImplementedError("Writers must implement a write method.")
+
+    def log_hyperparams(self, hyperparams: dict) -> None:
+        """
+        Logs hyperparameters.
+
+        Args:
+            hyperparams (dict): A dictionary of hyperparameters to log.
+        """
+        raise NotImplementedError("Writers must implement a log_hyperparams method.")
+
+    def plot(
+        self,
+        model: Module,
+        iteration: int,
+        plotting_functions: tuple[PlottingFunction, ...],
+    ) -> None:
+        """
+        Logs plots of the model using provided plotting functions.
+
+        Args:
+            model (Module): The model to plot.
+            iteration (int): The current iteration number.
+            plotting_functions (tuple[PlottingFunction, ...]): Functions used to
+                generate plots.
+        """
+        raise NotImplementedError("Writers must implement a plot method.")
+
+    def log_model(
+        self,
+        model: Module,
+        dataloader: Union[None, DataLoader, DictDataLoader],
+    ) -> None:
+        """
+        Logs the model and associated data.
+
+        Args:
+            model (Module): The model to log.
+            dataloader (DataLoader | DictDataLoader | None): DataLoader to use
+                for model input.
+        """
+        raise NotImplementedError("Writers must implement a log_model method.")
+
+    def print_metrics(self, result: OptimizeResult) -> None:
+        """Prints the metrics and loss in a readable format.
+
+        Args:
+            result (OptimizeResult): The optimization results to display.
+        """
+
+        # Find the key in result.metrics that contains "loss" (case-insensitive)
+        loss_key = next((k for k in result.metrics if "loss" in k.lower()), None)
+        if loss_key:
+            loss_value = result.metrics[loss_key]
+            msg = f"Iteration {result.iteration: >7} | {loss_key.title()}: {loss_value:.7f} -"
+        else:
+            msg = f"Iteration {result.iteration: >7} | Loss: None -"
+        msg += " ".join([f"{k}: {v:.7f}" for k, v in result.metrics.items() if k != loss_key])
+        print(msg)
+
+
+class TensorBoardWriter(BaseWriter):
+    """Writer for logging to TensorBoard.
+
+    Attributes:
+        writer (SummaryWriter): The TensorBoard SummaryWriter instance.
+    """
+
+    def __init__(self) -> None:
+        self.writer = None
+
+    def open(self, config: TrainConfig, iteration: int = None) -> SummaryWriter:
+        """
+        Opens the TensorBoard writer.
+
+        Args:
+            config: Configuration object containing settings for logging.
+            iteration (int, optional): The iteration step to start logging from.
+                Defaults to None.
+
+        Returns:
+            SummaryWriter: The initialized TensorBoard writer.
+        """
+        log_dir = str(config._log_folder)
```

Comment:
If `_log_folder` is accessed outside of `TrainConfig`, perhaps it shouldn't be a protected variable.

Reply:
We can make it public. My rational to keep log_folder and subfolder private in the TrainConfig was - the user will not be able to set these values - they will be for internal use, and will be updated internally by the configmanager. All other attributes of the TrainConfig can be modified by the user, but not these. What do you think?

Reply:
Sure. My suggestion would be to use a read-only mechanism.
```python
class TrainConfig:
    def __init__(..., log_folder,...)
        self._log_folder = log_folder

    @property
    def log_folder(self):
        return self.log_folder
``` 
This makes `log_folder` readable but not settable.

Reply:
Alright, here the proposed solution, I have added this to the docs as well. 
The user can specify three parameters for logging now. 

`root_folder`,  `create_subfolder_per_run`, `log_folder`.

- The `root_folder` and `create_subfolder_per_run` will work as expected, where logs will be saved inside a subfolder inside root folder. 

- The user can also specify a  `log_folder` (which will only work when `create_subfolder_per_run` = False (default)) - in which case the logs will be saved inside the log_folder. 

The additional `log_folder` argument might be useful in case someone wants load checkpoints and continue training from a specific run.